cmake/opentitan.cmake

```diff
@@ -0,0 +1,104 @@
+# Copyright 2024 ETH Zurich and University of Bologna.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Viviane Potocnik <vivianep@iis.ee.ethz.ch>
+
+include(ExternalProject)
+
+set(OPENTITAN_DIR ${CMAKE_CURRENT_BINARY_DIR}/../drivers/opentitan)
+
+if(EXISTS ${OPENTITAN_DIR})
+  file(REMOVE_RECURSE ${OPENTITAN_DIR})
+endif()
+
+set(SPARSE_CHECKOUT_FILE ${CMAKE_CURRENT_BINARY_DIR}/sparse-checkout)
+
+set(SPARSE_PATTERNS
+  "util/*"
+  "hw/ip/*"
+  "sw/device/lib/base/*"
+  "sw/device/lib/base/internal/*"
+  "sw/device/lib/dif/*"
+  "sw/device/lib/dif/autogen/*"
+)
+
+file(WRITE ${SPARSE_CHECKOUT_FILE} "")
+foreach(PATTERN ${SPARSE_PATTERNS})
+  file(APPEND ${SPARSE_CHECKOUT_FILE} "${PATTERN}\n")
+endforeach()
+
+# Add OpenTitan as an external project
+ExternalProject_Add(
+  opentitan
+  DOWNLOAD_COMMAND
+    git clone --no-checkout https://github.com/lowRISC/opentitan.git ${OPENTITAN_DIR}
+    COMMAND git -C ${OPENTITAN_DIR} config core.sparseCheckout true
+    COMMAND ${CMAKE_COMMAND} -E copy ${SPARSE_CHECKOUT_FILE} ${OPENTITAN_DIR}/.git/info/sparse-checkout
+    COMMAND git -C ${OPENTITAN_DIR} checkout master
+  UPDATE_COMMAND ""  # Disable updates to avoid unexpected changes
+  CONFIGURE_COMMAND ""  # No configure step needed
+  BUILD_COMMAND ""  # No build step needed
+  INSTALL_COMMAND ""  # No install step needed
+  SOURCE_DIR ${OPENTITAN_DIR}
+)
+
+message(STATUS "OpenTitan added as an external project.")
+
+set(OPENTITAN_SW_DIR ${OPENTITAN_DIR}/sw)
+set(OPENTITAN_HW_IP_DIR ${OPENTITAN_DIR}/hw/ip)
+set(OPENTITAN_UTIL_DIR ${OPENTITAN_DIR}/util)
+
+set(REGTOOL_PY ${OPENTITAN_UTIL_DIR}/regtool.py)
+set(UART_HJSON ${OPENTITAN_HW_IP_DIR}/uart/data/uart.hjson)
+set(UART_REGS_H ${OPENTITAN_SW_DIR}/device/lib/dif/autogen/uart_regs.h)
+
+add_custom_command(
+  OUTPUT ${UART_REGS_H}
+  COMMAND ${Python3_EXECUTABLE} ${REGTOOL_PY} ${UART_HJSON} -D > ${UART_REGS_H}
+  DEPENDS ${REGTOOL_PY} ${UART_HJSON}
+  COMMENT "Generating uart_regs.h"
+)
+
+add_custom_target(
+  generate_uart_regs ALL
+  DEPENDS ${UART_REGS_H}
+)
+
+add_dependencies(generate_uart_regs opentitan)
+
+ExternalProject_Get_Property(opentitan SOURCE_DIR)
+
+file(GLOB_RECURSE OPENTITAN_SOURCES
+    "${OPENTITAN_SW_DIR}/device/lib/base/*.c"
+    "${OPENTITAN_SW_DIR}/device/lib/dif/*.c"
+    "${OPENTITAN_SW_DIR}/device/lib/dif/autogen/*.c"
+    "${OPENTITAN_SW_DIR}/device/lib/runtime/*.c"
+)
+
+list(FILTER OPENTITAN_SOURCES EXCLUDE REGEX ".*crc32.*")
+list(FILTER OPENTITAN_SOURCES EXCLUDE REGEX ".*hardened.*")
```

Comment:
There is some brute-forcing going on here. What's up with `crc32` and `hardened`?

Reply:
I’ve noticed that the `opentitan` repository includes a considerable number of files that are either unnecessary or unsupported on our platforms. As you might recall from my discussion with Pirmin from OpenTitan, he suggested using the `OT_PLATFORM_RV32` flag for 32-bit RISC-V platforms. However, this flag also enables `crc32` instructions (cyclic redundancy check), which are part of the Bitmanip extensions that our toolchain does not support.

Given this, I don’t see the benefit of including the entire repository at this stage. Doing so introduces some drawbacks, such as:

1. Slowing down the cloning process and the building of opentitan_lib.
2. Including files with instructions that we currently cannot support.

For now, I believe it makes sense to keep the repository to a minimal set of essential files. As we progress, we can always revisit and add more files if the need arises.

Reply:
Definitely a discussion worthy of a comment - if I see these stray lines of filtering I start to wonder :)

Point 1, however, *should* be mute, as built files should not be rebuilt (read: I would set up an installation and reuse things that can be reused). 
Point 2 is fair enough. Adding it here is a decent choice imo, and I am a bit dubious whether this approach will scale to targets with different peripheral configurations - the drivers to be used are target-specific after all.